### PR TITLE
EIP-1898 implementation

### PR DESCRIPTION
**Pull request by badangrycloud**

### CODE REVIEW

File: rpc/eth/api.go [hunk 1]
```diff
@@ -25,9 +25,9 @@ import (
 	"github.com/oasisprotocol/oasis-sdk/client-sdk/go/modules/evm"
 	"github.com/oasisprotocol/oasis-sdk/client-sdk/go/types"
 
-	"github.com/oasisprotocol/emerald-web3-gateway/conf"
-	"github.com/oasisprotocol/emerald-web3-gateway/indexer"
-	"github.com/oasisprotocol/emerald-web3-gateway/rpc/utils"
+	"github.com/oasisprotocol/oasis-evm-web3-gateway/conf"
```
Comment (ptrus) [inline rpc/eth/api.go]: Please revert these import renames

File: rpc/eth/api.go [hunk 2]
```diff
@@ -167,11 +167,11 @@ func (api *PublicAPI) GetBlockTransactionCountByNumber(blockNum ethrpc.BlockNumb
 	return hexutil.Uint(n), nil
 }
 
-func (api *PublicAPI) GetStorageAt(address common.Address, position hexutil.Big, blockNum ethrpc.BlockNumber) (hexutil.Big, error) {
-	logger := api.Logger.With("method", "eth_getStorageAt", "address", address, "position", position, "block_number", blockNum)
+func (api *PublicAPI) GetStorageAt(address common.Address, position hexutil.Big, blockNrOrHash ethrpc.BlockNumberOrHash) (hexutil.Big, error) {
+	logger := api.Logger.With("method", "eth_getStorageAt", "address", address, "position", position, "block number or hash", blockNrOrHash)
```
Comment (ptrus) [inline rpc/eth/api.go]:
```suggestion
	logger := api.Logger.With("method", "eth_getStorageAt", "address", address, "position", position, "block_or_hash", blockNrOrHash)
```

Same in other places.

File: rpc/eth/api.go [hunk 3]
```diff
@@ -673,3 +674,19 @@ func (api *PublicAPI) Hashrate() hexutil.Uint64 {
...
 	return 0
 }
+
+// getBlockRound returns the block round from BlockNumberOrHash.
+func (api *PublicAPI) getBlockRound(logger *logging.Logger, blockNrOrHash ethrpc.BlockNumberOrHash) (uint64, error) {
+	switch {
+	case blockNrOrHash.BlockHash == nil && blockNrOrHash.BlockNumber == nil:
+		return 0, fmt.Errorf("types BlockHash and BlockNumber cannot be both nil")
+	case blockNrOrHash.BlockHash != nil:
+		round, err := api.backend.QueryBlockRound(*blockNrOrHash.BlockHash)
```
Comment (ptrus) [inline rpc/eth/api.go]:
```suggestion
		return api.backend.QueryBlockRound(*blockNrOrHash.BlockHash)
```

File: rpc/eth/api.go [hunk 4]
```diff
@@ -673,3 +674,19 @@ func (api *PublicAPI) Hashrate() hexutil.Uint64 {
 	logger.Debug("request")
 	return 0
 }
+
+// getBlockRound returns the block round from BlockNumberOrHash.
+func (api *PublicAPI) getBlockRound(logger *logging.Logger, blockNrOrHash ethrpc.BlockNumberOrHash) (uint64, error) {
+	switch {
+	case blockNrOrHash.BlockHash == nil && blockNrOrHash.BlockNumber == nil:
+		return 0, fmt.Errorf("types BlockHash and BlockNumber cannot be both nil")
+	case blockNrOrHash.BlockHash != nil:
```
Comment (ptrus) [inline rpc/eth/api.go]: According to the spec an error should be returned in case both `blockHash` and `blockNumber` are specified https://github.com/ethereum/EIPs/issues/2247#issuecomment-523167874

However, `ethrpc.BlockNumberOrHash` already handles that case when unmarshaling (https://github.com/ethereum/go-ethereum/blob/6ce4670bc03602bb6638c490f892e848647c15c0/rpc/types.go#L134-L136), so a separate check is not needed here at the moment. Maybe just add a comment that the case is already handled by `BlockNumberOrHash`.

File: rpc/eth/api.go [hunk 5]
```diff
@@ -673,3 +674,19 @@ func (api *PublicAPI) Hashrate() hexutil.Uint64 {
...
+		round, err := api.backend.QueryBlockRound(*blockNrOrHash.BlockHash)
+		return round, err
+	case blockNrOrHash.BlockNumber != nil:
+		round, err := api.roundParamFromBlockNum(logger, *blockNrOrHash.BlockNumber)
```
Comment (ptrus) [inline rpc/eth/api.go]:
```suggestion
		return api.roundParamFromBlockNum(logger, *blockNrOrHash.BlockNumber)
```

File: rpc/eth/api.go [hunk 6]
```diff
@@ -599,7 +600,7 @@ func (api *PublicAPI) GetLogs(filter filters.FilterCriteria) ([]*ethtypes.Log, e
 	filtered := make([]*ethtypes.Log, 0, len(ethLogs))
 	for _, log := range ethLogs {
 		// Filter by address.
-		addressMatch := len(filter.Addresses) == 0
+		addressMatch := len(filtered) == 0
```
Comment (ptrus) [inline rpc/eth/api.go]: Revert this change

File: rpc/eth/api.go [hunk 7]
```diff
@@ -673,3 +674,19 @@ func (api *PublicAPI) Hashrate() hexutil.Uint64 {
 	logger.Debug("request")
 	return 0
 }
+
+// getBlockRound returns the block round from BlockNumberOrHash.
+func (api *PublicAPI) getBlockRound(logger *logging.Logger, blockNrOrHash ethrpc.BlockNumberOrHash) (uint64, error) {
```
Comment (ptrus) [inline rpc/eth/api.go]: This function doesn't use the `logger *logging.Logger` so remove the argument.

**Review by ptrus — COMMENTED** [review]
Thanks, looks mostly good, please address the comments.

File: rpc/eth/api.go (hunk 7 above)
Comment (badangrycloud) [reply rpc/eth/api.go:679]: this is use in the roundParamFromBlockNum func so is needed to that function it's called inside

**Review by ptrus — APPROVED** [review]
Looks good now. 

Two more things before this can get merged:
- squash commits in a single commit titled: `feat: implement EIP-1898`
- rebase on `main` branch
- 
Thanks!
